### ISSUE TITLE
Add NinePatch Support (i.e. Stretchable Sprites for Menus)

### ISSUE THREAD

**Issue by Nodragem** [BabylonJS/Babylon.js#5315]
To create menus, it can be very useful to have ninepatch image support, such as seen in Android development or in libGDX:
https://www.dev2qa.com/android-nine-patch-image-example/
https://github.com/libgdx/libgdx/wiki/Ninepatches

In Unity, they made they own soup and called it 9-slicing sprites:
https://docs.unity3d.com/Manual/9SliceSprites.html

Either way, it is useful :)

**Comment by Nodragem** [comment]
I'm just wondering, is 4.0 the next release? no 3.4?

**Comment by deltakosh** [comment]
4.0 is the next one

**Comment by deltakosh** [comment]
So to define the 9patches, we need to:
- create a new NinePatchImage inheriting from Image
- add 4 properties: SliceLeft, SliceRight, SliceTop and SliceBottom which will define the borders

Behavior would be like Unity Sliced mode

**Comment by Nodragem** [comment]
I am thinking that:
- we don't necessarily have an editor like Unity where to edit the 9-Slice Sprite.
- to provide 4 corner properties in the function call sounds like we are using magic numbers / storing data in our code.

Hence I think that the solution adopted by Android SDK and libGDX may be best.
The user defined the stretchable part of the image in the image itself. No need of a 9-Slice Sprite editor, MS Paint is enough and the data is kept outside the code. 
See https://github.com/libgdx/libgdx/wiki/Ninepatches

**Comment by deltakosh** [comment]
Well not sure this is what I want as a dev. I definitely prefer defining the size as numbers: it is precise, I can control it or animate it, it is faster ad the app does not need to read hacky values stored in the file

**Comment by Nodragem** [comment]
I think there are few arguments that can be found against your points:

> - hacky values?

 
On the image row/column that defines your borders, the pixel values for the transparent channel will look like that:

0, 0, 0, 0, 0, 255, 255, 255, 255, 255, 0, 0, 0, 0

I think we can precisely find the index of the first and last 255.

> - faster:

That is a one time operation, to read 1 or 2 lines of pixel should be virtually instantaneous. 
LibGDX was famous for its very good performance; and Android does not seem to suffer from the performance of its 9-patches.

> - I definitely prefer defining the size as numbers: it is precise, I can control it or animate it

- **precise:** As argued above, the precision does not seem to be a problem. 
- **control and animation:** I can think of 2 arguments, 1) it seems to me that there is no practical use of animating where your corners should be in your button image; 2) it is not because you initialised your image with `bottom_corner = 129` that you can't animate/control the property `bottom_corner` afterwards; same thing applies when the property is initialised from a 9-patch image.

Finally, in term of usability, I would prefer not to have to count 129 pixels on my screen to write that value in my code, and then start again for the other corners. Yes, you could use an advanced 2D image editor (e.g. GIMP) to add rulers/guides on your image to count your pixels, but each time you change your mind you need to 1) move your rulers in GIMP, 2) read and write the new values in your code. In comparison, with the 9-patch image, you can just open-up a simple 2D tool (e.g. MS Paint), draw a line and save.

We may need someone else to give their opinion (anybody?), but I am happy to follow whatever you decide.

**Comment by deltakosh** [comment]
I disagree but you are right let's ask for more thoughts. Do you mind creating a post on the forum for that?

**Comment by Nodragem** [comment]
will do (having some trouble with my working station right now).

**Comment by Nodragem** [comment]
Not that much feedback on the forum. But in case Github fellows missed my new opinion, I will write it down here.

I start to think that both methods should be implemented as well as a third method.  Let me explain.

My thinking is that libGDX is a great game engine and they faced the same issues we will face when they implemented NinePatch. And I believe that to copy their solution will be a good place to start.

See: https://github.com/libgdx/libgdx/wiki/Ninepatches 

As I undertstand it from the link above, LibGDX provides three ways to create a NinePatch:

- **Method 1:** Set the stretchable part of the NinePatch in code
- **Method 2:** Set the stretchable parts of the NinePatch directly in the texture
- **Method 3:** Set the stretchable parts of the NinePatch from a TexturePack

As we also would like to implement a TexturePack support (see https://github.com/BabylonJS/Babylon.js/issues/5314), it is worth checking how libGDX dealt with NinePatch support in their Texture Packer.

Basically, if I understand well, the NinePatch support in TexturePack is dealt with this way in libGDX:

1) the user defines NinePatch directly in the texture using **Method 2**
2) when the NinePatch texture is packed in a TexturePack, the rows/columns that encode the stretching/padding information are removed and the information is placed in the JSON file of the TexturePack
3) when the NinePatch texture is imported from the TexturePack, **Method 1** is used to apply the stretching/padding information contained in the JSON at runtime 

Hence, I think that both methods (Method 1 and Method 2) could be supported by BabylonJS for now.

Any thoughts?

**Comment by deltakosh** [comment]
What we could do is supporting 1 and 2
We can get the value for 1 from 2..Thoughts?

**Comment by deltakosh** [comment]
So, todo:
- [ ] Add a new NinePatchImage inheriting from image
- [ ] Implement option 1
- [ ] Implement option 2 to read values and send it to 1
- [ ] Update demo here: https://www.babylonjs-playground.com/#HATGQZ

**Comment by deltakosh** [comment]
I can work on it unless you want to take a stab?

**Comment by Nodragem** [comment]
I would like too, but I realistically can't :/ I am starting a new contract tomorrow (using BabylonJS, let's see how it goes!). I am still hoping to contribute somehow.

**Comment by deltakosh** [comment]
No worry I can take care of it :) I just wanted to be sure we are now developing twice :)

And good luck for your job (Merde like we say in french ;))

**Comment by Nodragem** [comment]
For your tests:
[panel_blue2x.9.zip](https://github.com/BabylonJS/Babylon.js/files/2489352/panel_blue2x.9.zip)

This is the intended results:
![opera_2018-10-17_22-23-35](https://user-images.githubusercontent.com/10520249/47117236-bed09e00-d25b-11e8-82a8-59cce85a97f8.png)

You can make ninepatches from pretty any images here: http://romannurik.github.io/AndroidAssetStudio/nine-patches.html#&sourceDensity=320&name=example

**Comment by Nodragem** [comment]
In fact, I am French too ^^ Merci!

**Comment by deltakosh** [comment]
thanks a lot!

**Comment by deltakosh** [comment]
Demo here:
https://www.babylonjs-playground.com/#G5H9IN#2

(will work as soon as nightly is compiled and merged)

**Comment by Nodragem** [comment]
very nice, thank you for this feature!

**Comment by deltakosh** [comment]
My pleasure:)
